internal/buffer/buffer.go

```diff
@@ -1140,7 +1141,41 @@ var BracePairs = [][2]rune{
 	{'[', ']'},
 }
 
+func (b *Buffer) getSortedSyntaxIndices(lineN int) []int {
+	keys := make([]int, 0, len(b.Match(lineN)))
+	for k := range b.Match(lineN) {
+		keys = append(keys, k)
+	}
+	sort.Ints(keys)
+	return keys
+}
+
+// Returns the Group (syntax highlight group ID) at the specified location and a boolean
+// that indicates if a group is found or not
+func (b *Buffer) GetGroupAtLoc(loc Loc) (highlight.Group, bool) {
+	sortedIndices := b.getSortedSyntaxIndices(loc.Y)
+	i := sort.SearchInts(sortedIndices, loc.X)
+	if i == 0 || i == len(sortedIndices) {
+		return 0, false
+	}
+	if sortedIndices[i] == loc.X && b.Match(loc.Y)[sortedIndices[i]] != 0 {
+		return b.Match(loc.Y)[sortedIndices[i]], true
+	}
+	return b.Match(loc.Y)[sortedIndices[i - 1]], true
+}
```

Comment:
Maybe I'm missing something but is the sorting necessary?
```suggestion
func (b *Buffer) GetGroupAtLoc(loc Loc) (highlight.Group, bool) {
	found := -1
	for i := range b.Match(loc.Y) {
		if i > found && i <= loc.X {
			found = i
		}
	}
	if found == -1 {
		return 0, false
	}
	return b.Match(loc.Y)[found], true
}
```

Reply:
The suggested change won't work when the cursor is at the end of a syntax change, which has an empty group and needs the previous group id. 

The sorted indices are now being reused for the rest of the lines without iterating the whole line again, which will be useful for extreme cases where the line is very long, for example, minified js or json.

Reply:
> The suggested change won't work when the cursor is at the end of a syntax change, which has an empty group and needs the previous group id.

Can you provide an example of this / how to test it? It should be fixable – if the only reason you need the sorted indices is to call `sort.SearchInts` afterwards you can always do it with a simple loop (less code and no allocations).